src/program/mod.rs

```diff
@@ -93,12 +95,12 @@ impl Program {
     /// Expand any instructions in the program which have a matching calibration, leaving the others
     /// unchanged. Recurses though each instruction while ensuring there is no cycle in the expansion
     /// graph (i.e. no calibration expands directly or indirectly into itself)
-    pub fn expand_calibrations(&self) -> error::ProgramResult<Self> {
+    pub fn expand_calibrations(&self) -> Result<Self> {
         let mut expanded_instructions: Vec<Instruction> = vec![];
 
         // TODO: Do this more efficiently, possibly with Vec::splice
         for instruction in &self.instructions {
-            match self.calibrations.expand(instruction, &[])? {
+            match convert_leftover(self.calibrations.expand(instruction, &[]))? {
```

Comment:
parsing-related utilities seem to be leaking out of parsing-related code here

Reply:
This may generally be parsing related, but the function itself is defined in `src/program/error/result.rs`, which is not a parsing-specific module. There isn't really a better way to handle this, IMO. `expand()` returns an error of `ProgramError<Option<Vec<Instruction>>>`, while `expand_calibrations()` returns `ProgramError<Program>`.

I've made a commit that is the only other fix to this that I can think of -- changing the error to `ProgramError<Program>`. I initially avoided this because `expand()` is `pub fn` and so, as far as I can tell, part of the public API. As such, making the error `ProgramError<Program>` feels off, even if it works in the type system.

This commit works because the only callers of `expand()` are `expand()` itself and `expand_calibrations()`.

Reply:
This change might actually be preferable, because it's the only use of `convert_leftover`, the removal of which made moving `Leftover` from `ProgramError` to `SyntaxError` easier to do.